core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java

```diff
@@ -297,6 +330,14 @@ public static long rescale(long value, int fromScale, int toScale)
         if (toScale < fromScale) {
             throw new IllegalArgumentException("target scale must be larger than source scale");
         }
+
```

Comment:
It would be clearer when you check that `toScale <= MAX_SHORT_PRECISION` first.
Then you can just make sure that `value * longTenToNth(toScale - fromScale)` doesn't overtflow.